### PR TITLE
remove_empty removes empty strings

### PR DESCRIPTION
**Pull request by strengejacke**
Variables/columns are also removed if these are of type character and only contain `""`.

### CODE REVIEW

**Comment by strengejacke** [comment]
@IndrajeetPatil @etiennebacher Would you say this new argument makes sense? And what about a potential name for that arg?

**Comment by etiennebacher** [comment]
I think it's not necessary to have an additional argument, I don't see a context where someone would want to keep only `""`.  And if they want to remove only columns with `NA`s, the code to do it is not very complicated

File: R/remove_empty.R [hunk 1]
```diff
@@ -42,11 +44,19 @@
...
 #' @export
-empty_columns <- function(x) {
+empty_columns <- function(x, include_empty_string = TRUE) {
   if ((!is.matrix(x) && !is.data.frame(x)) || ncol(x) < 2) {
     vector("numeric")
   } else {
-    which(colSums(is.na(x)) == nrow(x))
+    all_na <- colSums(is.na(x)) == nrow(x)
+    if (include_empty_string) {
+      all_empty <- sapply(x, function(i) {
```
Comment (etiennebacher) [inline R/remove_empty.R:53]: you can use `vapply()` instead to save a few microseconds 😄

Comment (strengejacke) [reply R/remove_empty.R:53]: oh yes, `vapply` and `tapply` receive too little love from R coders ;-)

**Comment by strengejacke** [comment]
I'm not sure about the `FUN.VALUE` argument, I always get it to work somehow, but not sure if it's correctly used?

``` r
x <- data.frame(
  a = c(1, 2, 3, NA, 5),
  b = c("", NA, "", NA, ""),
  c = c(NA, NA, NA, NA, NA),
  d = c(1, NA, 3, NA, 5),
  e = c("", "", "", "", ""),
  f = factor(c("", "", "", "", "")),
  stringsAsFactors = FALSE
)

all_empty <- vapply(x, function(i) {
  is.character(i) && max(nchar(i), na.rm = TRUE) == 0
}, FUN.VALUE = TRUE)
all_empty
#>     a     b     c     d     e     f 
#> FALSE  TRUE FALSE FALSE  TRUE FALSE


all_empty <- sapply(x, function(i) {
  is.character(i) && max(nchar(i), na.rm = TRUE) == 0
})
all_empty
#>     a     b     c     d     e     f 
#> FALSE  TRUE FALSE FALSE  TRUE FALSE
```

<sup>Created on 2022-09-28 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Comment by etiennebacher** [comment]
> I'm not sure about the FUN.VALUE argument, I always get it to work somehow, but not sure if it's correctly used?

It ensures that all output values are of the same class. I like [this example](https://www.educative.io/answers/how-to-use-the-vapply-function-in-r):
``` r
A <- c(-1: -10)
B <- c(1:5)
C <- c("x", "y", "z")
lst <- list(A,B,C)

sapply(lst, min)
#> [1] "-10" "1"   "x"
vapply(lst, min, FUN.VALUE = numeric(1))
#> Error in vapply(lst, min, FUN.VALUE = numeric(1)): les valeurs doivent être de type 'double',
#>  mais FUN(X[[3]]) est de type 'character'
```
So I think the "proper use" in our case should be this:
``` r
vapply(iris, is.numeric, logical(1))
#> Sepal.Length  Sepal.Width Petal.Length  Petal.Width      Species 
#>         TRUE         TRUE         TRUE         TRUE        FALSE
```
but I don't know why using `TRUE` instead of `logical(1)` works.

**Comment by strengejacke** [comment]
@IndrajeetPatil @etiennebacher 

Should we also include "empty" factors that only have `""` values? Currently, this is only checked for characters:

``` r
tmp <- data.frame(
  a = c(1, 2, 3, NA, 5),
  b = c("", NA, "", NA, ""),
  c = c(NA, NA, NA, NA, NA),
  d = c(1, NA, 3, NA, 5),
  e = c("", "", "", "", ""),
  f = factor(c("", "", "", "", "")),
  g = factor(c("", NA, "", NA, "")),
  stringsAsFactors = FALSE
)

tmp
#>    a    b  c  d e f    g
#> 1  1      NA  1         
#> 2  2 <NA> NA NA     <NA>
#> 3  3      NA  3         
#> 4 NA <NA> NA NA     <NA>
#> 5  5      NA  5

datawizard::empty_columns(tmp)
#> b c e 
#> 2 3 5
```

<sup>Created on 2022-09-28 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Comment by etiennebacher** [comment]
> Should we also include "empty" factors that only have "" values? 

I would say yes

**Comment by strengejacke** [comment]
@IndrajeetPatil @etiennebacher 

# Old behaviour

``` r
tmp <- data.frame(
  a = c(1, 2, 3, NA, 5),
  b = c("", NA, "", NA, ""),
  c = c(NA, NA, NA, NA, NA),
  d = c(1, NA, 3, NA, 5),
  e = c("", "", "", "", ""),
  f = factor(c("", "", "", "", "")),
  g = factor(c("", NA, "", NA, "")),
  stringsAsFactors = FALSE
)

tibble::as_tibble(tmp)
#> # A tibble: 5 × 7
#>       a b     c         d e     f     g    
#>   <dbl> <chr> <lgl> <dbl> <chr> <fct> <fct>
#> 1     1 ""    NA        1 ""    ""    ""   
#> 2     2 <NA>  NA       NA ""    ""    <NA> 
#> 3     3 ""    NA        3 ""    ""    ""   
#> 4    NA <NA>  NA       NA ""    ""    <NA> 
#> 5     5 ""    NA        5 ""    ""    ""

datawizard::empty_columns(tmp)
#> c 
#> 3

datawizard::empty_rows(tmp)
#> integer(0)
```

# New behaviour

``` r
tmp <- data.frame(
  a = c(1, 2, 3, NA, 5),
  b = c("", NA, "", NA, ""),
  c = c(NA, NA, NA, NA, NA),
  d = c(1, NA, 3, NA, 5),
  e = c("", "", "", "", ""),
  f = factor(c("", "", "", "", "")),
  g = factor(c("", NA, "", NA, "")),
  stringsAsFactors = FALSE
)

tibble::as_tibble(tmp)
#> # A tibble: 5 × 7
#>       a b     c         d e     f     g    
#>   <dbl> <chr> <lgl> <dbl> <chr> <fct> <fct>
#> 1     1 ""    NA        1 ""    ""    ""   
#> 2     2 <NA>  NA       NA ""    ""    <NA> 
#> 3     3 ""    NA        3 ""    ""    ""   
#> 4    NA <NA>  NA       NA ""    ""    <NA> 
#> 5     5 ""    NA        5 ""    ""    ""

datawizard::empty_columns(tmp)
#> b c e f g 
#> 2 3 5 6 7

datawizard::empty_rows(tmp)
#> [1] 4
```

I think this PR is ready to merge, if you're ok with these changes.

**Comment by IndrajeetPatil** [comment]
Sorry, didn't follow the whole discussion, but the output looks good to me:

``` r
library(datawizard)

tmp <- dplyr::tibble(
  a = c(1, 2, 3, NA, 5),
  b = c("", NA, "", NA, ""),
  c = c(NA, NA, NA, NA, NA),
  d = c(1, NA, 3, NA, 5),
  e = c("", "", "", "", ""),
  f = factor(c("", "", "", "", "")),
  g = factor(c("", NA, "", NA, ""))
)

remove_empty_columns(tmp)
#> # A tibble: 5 × 2
#>       a     d
#>   <dbl> <dbl>
#> 1     1     1
#> 2     2    NA
#> 3     3     3
#> 4    NA    NA
#> 5     5     5

remove_empty_rows(tmp)
#> # A tibble: 4 × 7
#>       a b     c         d e     f     g    
#> * <dbl> <chr> <lgl> <dbl> <chr> <fct> <fct>
#> 1     1 ""    NA        1 ""    ""    ""   
#> 2     2 <NA>  NA       NA ""    ""    <NA> 
#> 3     3 ""    NA        3 ""    ""    ""   
#> 4     5 ""    NA        5 ""    ""    ""

remove_empty(tmp)
#> # A tibble: 4 × 2
#>       a     d
#>   <dbl> <dbl>
#> 1     1     1
#> 2     2    NA
#> 3     3     3
#> 4     5     5
```

<sup>Created on 2022-09-29 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Review by IndrajeetPatil — APPROVED** [review]

**Pull request merged**